### ISSUE TITLE
Runtime application hung in Process>>shutdown

### ISSUE THREAD

**Issue by jgfoster** [dolphinsmalltalk/Dolphin#353]
When I interrupt it with [Ctrl]+[F12] the stack includes the following:

{1A040154: cf 1A040139, sp 1A040168, bp 1A040154, ip 13, RaisedSignal(Exception)>>signal}
	receiver: a RaisedSignal

{1A040138: cf 1A04011D, sp 1A04014C, bp 1A040138, ip 3, Signal>>signal}
	receiver: a Signal

{1A04011C: cf 1A040101, sp 1A040130, bp 1A04011C, ip 4, PCSessionManager(SessionManager)>>onUserBreak}
	receiver: a PCSessionManager

{1A040100: cf 1A0400E1, sp 1A040114, bp 1A0400FC, ip 6, ProcessorScheduler>>userBreak:}
	receiver: a ProcessorScheduler
	arg[0]: nil

{1A0400E0: cf 1A0400B5, sp 1A0400F4, bp 1A0400D8, ip 16, [] in ProcessorScheduler>>vmi:list:no:with:}
	receiver: a ProcessorScheduler

{1A0400B4: cf 1A040095, sp 1A0400C8, bp 1A0400B0, ip 18, BlockClosure>>ifCurtailed:}
	receiver: [] @ 0 in nil
	arg[0]: [] @ 24 in ProcessorScheduler>>vmi:list:no:with:

{1A040094: cf 1A040065, sp 1A0400A8, bp 1A040084, ip 27, ProcessorScheduler>>vmi:list:no:with:}
	receiver: a ProcessorScheduler
	arg[0]: 96
	arg[1]: a LinkedList
	arg[2]: 11
	arg[3]: nil

{1A040064: cf 1A040049, sp 1A04007C, bp 1A040064, ip 16, Process>>shutdown}
	receiver: a Process('Main' base 1A040000 [ACTIVE] in SessionManager>>logError: sp=00000000 ip=4 list=01770010)

{1A040048: cf 00000001, sp 1A04005C, bp 1A040048, ip 21, [] in BlockClosure>>newProcess}
	receiver: [] @ 10 in InputState>>forkMain

<Bottom of stack>

**Comment by blairmcg** [comment]
This sounds like the known VM issue where if an attempt is made to shutdown inside a callback that this fails and leaves the process running in a zombie state. The quit primitive works by throwing a Win32 exception that is trapped down in the main entry point function of the VM. This is done to unwind the stack cleanly running any clean-up blocks on the way. Or at least that is the theory. When Dolphin was originally written this worked fine because Windows allowed exceptions to be thrown over user->kernel->user mode transitions and be caught in the user mode code on the other side. Since Windows 7 (I think) it hasn't allowed this, so if Dolphin is (for example) inside a callback to process a windows message, then the exit attempt fails.
One sometimes encountered this when exiting a normal development image. I don't know of any good solution at present.

**Comment by jgfoster** [comment]
When you refer to 'shutdown' is that an Image shutdown or simply a Process shutdown?

**Comment by blairmcg** [comment]
Image shutdown

**Comment by jgfoster** [comment]
Hmm... I observed this during regular operations and the code in BlockClosure>>newProcess sends
    Processor activeProcess shutdown
So I wasn't thinking of Image shutdown.

**Comment by blairmcg** [comment]
In that case it's probably waiting on the pending terminations list. This is where the VM puts Smalltalk processes sent #terminate to wait for the associated overlapped call thread to terminate. It terminates the overlapped call thread (if any) by sending it an APC (Asynchronous Procedure Call). These can only be delivered when the thread enters an alertable wait state. The reason for the dance is to raise an exception in the thread so that any unwind handlers on the stack are run in the hope that this will allow it to terminate cleanly. Just terminating the thread is not a good idea, as explained on the [MSDN page for TerminateThread ](https://msdn.microsoft.com/en-us/library/windows/desktop/ms686717(v=vs.85).aspx) in the paragraph that begins "TerminateThread is a dangerous function that should only be used in the most extreme cases."
You could find out what is going on by attaching the visual studio debugger and looking at what the threads are doing. You should be able to work out which of the thread is the overlapped call thread for the blocked Process by looking at the m_thread field of the Process - this is a pointer to a C++ OverlappedCall object encoded as a SmallInteger by setting the bottom bit. The OverlappedCall object has fields for the thread handle and thread id. The latter should help you relate back to the debugger thread list.
If it isn't due to overlapped call termination, then I'm out of ideas.

**Comment by fxgallego** [comment]
HI @blairmcg is this [bug report](https://groups.google.com/forum/#!searchin/comp.lang.smalltalk.dolphin/error$20on$20exit%7Csort:relevance/comp.lang.smalltalk.dolphin/OTMZDlzcG8s/VHRRyzgtJUoJ) related to this issue?
We are currently randomly experiencing this problem in a Windows Server 2012

**Comment by blairmcg** [comment]
It looks like the issue I described with exceptions being thrown across a kernel-user mode transition, yes. In the stack trace you can see that the quit primitive is being called inside a windows message handler (for WM_QUERYENDSESSION). That's not too surprising given the implementation of `SessionManager>>onQueryWindowsShutdown` and `SessionManager>>quit:`. You could try overriding one or the other so that the actual call to quit primitive is called from a deferred action. This should allow the window procedure to return before the terminate exception is thrown.

**Comment by fxgallego** [comment]
I don't know now if this issue is related with the original posted by @jgfoster, but for future reference I'm documenting here.

We did the following:

1) Reproduce the error in Windows Server 2012 and Windows 10 using the workspace:
Evaluate: `SessionManager current quit: 0`
Result:
Windows Server 2012 warns about Dolphin Smalltalk has stopped working
Windows 10 doesn't warns but an errors file is generated with the same stack contents as W2012
Stack:
```
********************************************************************************
************************** Dolphin Crash Dump Report ***************************
10:09:28 AM, 4/25/2017: Dolphin7.exe caused an unhandled Win32 Exception A0000204
at 73E7A6F2 in module 73DA0000 (C:\WINDOWS\System32\KERNELBASE.dll)
*----> Exception Parameters <----*
00000000    
*----> CPU Context for thread 0x2cb0 <----*
EAX = 00C0FA18 EBX = 08CD0654 ECX = 00000001
ESI = 08CD0674 EDI = 08F562C8 EIP = 73E7A6F2
ESP = 00C0FA18 EBP = 00C0FA74 EFL = 00200216
CS = 0023 SS = 002B DS = 002B
ES = 002B FS = 0053 GS = 002B
*----> VM Context <----*
Process: {08CD0004:size 695 words, suspended frame 08CD0AC5, priority 6, callbacks 1
last failure 0:nil, FPE mask 3, thread 00000000}
Active Method: 'SessionManager'>>quit:
IP: 08F562C8 (8)
SP: 08CD0674
BP: 08CD0654 (389)
ActiveFrame: {08CD0658: cf 08CD0639, sp 08CD066C, bp 08CD0654, ip 4, 'DevelopmentSessionManager'('SessionManager')>>quit:}
    receiver: a 'DevelopmentSessionManager'
    arg[0]: 0
New Method: 'SessionManager'>>primQuit:
Message Selector: primQuit:
*----> Stack Back Trace <----*
{08CD0658: cf 08CD0639, sp 08CD066C, bp 08CD0654, ip 4, 'DevelopmentSessionManager'('SessionManager')>>quit:}
    receiver: a 'DevelopmentSessionManager'
    arg[0]: 0
{08CD0638: cf 08CD061D, sp 08CD064C, bp 08CD0638, ip 5, 'UndefinedObject'>>**Non-symbol: 'doIt'**}
    receiver: nil
{08CD061C: cf 08CD05FD, sp 08CD0630, bp 08CD0618, ip 4, 'CompiledExpression'>>value:}
    receiver: a 'CompiledExpression'
    arg[0]: nil
{08CD05FC: cf 08CD05CD, sp 08CD0610, bp 08CD05E8, ip 28, 'SmalltalkWorkspace'>>evaluateRange:ifFail:debug:}
    receiver: a 'SmalltalkWorkspace'
    arg[0]: a 'Interval'
    arg[1]: [] @ 12 in 'SmalltalkWorkspace'>>evaluateItIfFail:debug:
    arg[2]: false
    stack temp[0]: a 'CompilationResult'
    stack temp[1]: nil
{08CD05CC: cf 08CD059D, sp 08CD05E0, bp 08CD05B8, ip 18, 'SmalltalkWorkspace'>>evaluateItIfFail:debug:}
    receiver: a 'SmalltalkWorkspace'
    arg[0]: [] @ 3 in a CompiledExpression
    arg[1]: false
    stack temp[0]: a 'Interval'
    stack temp[1]: nil
    stack temp[2]: nil
{08CD059C: cf 08CD057D, sp 08CD05B0, bp 08CD0598, ip 4, 'SmalltalkWorkspace'>>evaluateItIfFail:}
    receiver: a 'SmalltalkWorkspace'
    arg[0]: [] @ 3 in a CompiledExpression
{08CD057C: cf 08CD0561, sp 08CD0590, bp 08CD057C, ip 4, 'SmalltalkWorkspace'>>evaluateIt}
    receiver: a 'SmalltalkWorkspace'
{08CD0560: cf 08CD0541, sp 08CD0574, bp 08CD055C, ip 4, 'Symbol'>>forwardTo:}
    receiver: evaluateIt
    arg[0]: a 'SmalltalkWorkspace'
{08CD0540: cf 08CD0521, sp 08CD0554, bp 08CD053C, ip 4, 'CommandDescription'>>performAgainst:}
    receiver: a 'CommandDescription'
    arg[0]: a 'SmalltalkWorkspace'
{08CD0520: cf 08CD04FD, sp 08CD0534, bp 08CD0520, ip 14, [] in 'Command'>>value}
    receiver: a 'Command'
{08CD04FC: cf 08CD04DD, sp 08CD0510, bp 08CD04F8, ip 18, 'BlockClosure'>>ifCurtailed:}
    receiver: [] @ 0 in nil
    arg[0]: [] @ 17 in 'Command'>>value
{08CD04DC: cf 08CD04B9, sp 08CD04F0, bp 08CD04D4, ip 3, 'BlockClosure'>>ensure:}
    receiver: [] @ 11 in 'Command'>>value
    arg[0]: [] @ 17 in 'Command'>>value
    stack temp[0]: nil
{08CD04B8: cf 08CD049D, sp 08CD04CC, bp 08CD04B8, ip 22, 'Command'>>value}
    receiver: a 'Command'
{08CD049C: cf 08CD047D, sp 08CD04B0, bp 08CD0498, ip 30, 'ShellView'>>performCommand:}
    receiver: a 'ShellView'
    arg[0]: a 'Command'
{08CD047C: cf 08CD045D, sp 08CD0490, bp 08CD0478, ip 4, 'SmalltalkWorkspaceDocument'('Shell')>>performCommand:}
    receiver: a 'SmalltalkWorkspaceDocument'
    arg[0]: a 'Command'
{08CD045C: cf 08CD0435, sp 08CD0470, bp 08CD0450, ip 10, 'CommandQuery'>>perform}
    receiver: a 'CommandQuery'
    stack temp[0]: a 'Command'
    stack temp[1]: nil
    stack temp[2]: a 'SmalltalkWorkspaceDocument'
{08CD0434: cf 08CD0411, sp 08CD0448, bp 08CD042C, ip 8, 'DelegatingCommandPolicy'('CommandPolicy')>>route:}
    receiver: a 'DelegatingCommandPolicy'
    arg[0]: a 'CommandDescription'
    stack temp[0]: a 'CommandQuery'
{08CD0410: cf 08CD03E9, sp 08CD0424, bp 08CD040C, ip 15, [] in 'ShellView'('View')>>onCommand:}
    receiver: a 'ShellView'
{08CD03E8: cf 08CD03C9, sp 08CD03FC, bp 08CD03E4, ip 18, 'BlockClosure'>>ifCurtailed:}
    receiver: [] @ 0 in nil
    arg[0]: [] @ 18 in 'Cursor'>>showWhile:
{08CD03C8: cf 08CD03A5, sp 08CD03DC, bp 08CD03C0, ip 3, 'BlockClosure'>>ensure:}
    receiver: [] @ 11 in 'View'>>onCommand:
    arg[0]: [] @ 18 in 'Cursor'>>showWhile:
    stack temp[0]: nil
{08CD03A4: cf 08CD037D, sp 08CD03B8, bp 08CD0398, ip 27, 'Cursor'>>showWhile:}
    receiver: a 'Cursor'
    arg[0]: [] @ 11 in 'View'>>onCommand:
    stack temp[0]: nil
    stack temp[1]: a 'ExternalHandle'(00010003)
{08CD037C: cf 08CD0359, sp 08CD0390, bp 08CD0374, ip 17, 'ShellView'('View')>>onCommand:}
    receiver: a 'ShellView'
    arg[0]: a 'CommandDescription'
    stack temp[0]: nil
{08CD0358: cf 08CD0325, sp 08CD036C, bp 08CD0340, ip 19, 'ShellView'('View')>>wmCommand:wParam:lParam:}
    receiver: a 'ShellView'
    arg[0]: 273
    arg[1]: 70108
    arg[2]: 0
    stack temp[0]: 4572
    stack temp[1]: a 'CommandDescription'
    stack temp[2]: nil
{08CD0324: cf 08CD02F5, sp 08CD0338, bp 08CD0310, ip 23, 'ShellView'('View')>>dispatchMessage:wParam:lParam:}
    receiver: a 'ShellView'
    arg[0]: 273
    arg[1]: 70108
    arg[2]: 0
    stack temp[0]: nil
    stack temp[1]: wmCommand:wParam:lParam:
{08CD02F4: cf 08CD02C1, sp 08CD0308, bp 08CD02E4, ip 70, [] in 'InputState'>>wndProc:message:wParam:lParam:cookie:}
    receiver: nil
{08CD02C0: cf 08CD0299, sp 08CD02D4, bp 08CD02BC, ip 18, 'BlockClosure'>>ifCurtailed:}
    receiver: [] @ 0 in nil
    arg[0]: [] @ 11 in 'ProcessorScheduler'>>callback:evaluate:
{08CD0298: cf 08CD0275, sp 08CD02B4, bp 08CD0290, ip 15, 'ProcessorScheduler'>>callback:evaluate:}
    receiver: a 'ProcessorScheduler'
    arg[0]: 6323570
    arg[1]: [] @ 64 in 'InputState'>>wndProc:message:wParam:lParam:cookie:
{08CD0274: cf 08CD0231, sp 08CD0288, bp 08CD025C, ip 73, 'InputState'>>wndProc:message:wParam:lParam:cookie:}
    receiver: a 'InputState'
    arg[0]: 460818
    arg[1]: 273
    arg[2]: 70108
    arg[3]: 0
    arg[4]: 6323570
    stack temp[0]: a 'ShellView'
{08CD0230: cf 08CD020D, sp 08CD0254, bp 08CD0228, ip 17, 'ShellView'>>translateAccelerator:}
    receiver: a 'ShellView'
    arg[0]: a 'MSG'
    stack temp[0]: a 'ExternalHandle'(55170D73)
{08CD020C: cf 08CD01E9, sp 08CD0220, bp 08CD0204, ip 6, 'ShellView'>>preTranslateKeyboardInput:}
    receiver: a 'ShellView'
    arg[0]: a 'MSG'
    stack temp[0]: nil
{08CD01E8: cf 08CD01C9, sp 08CD01FC, bp 08CD01E4, ip 6, 'ShellView'('View')>>preTranslateMessage:}
    receiver: a 'ShellView'
    arg[0]: a 'MSG'
{08CD01C8: cf 08CD01A1, sp 08CD01DC, bp 08CD01BC, ip 17, 'InputState'>>preTranslateMessage:}
    receiver: a 'InputState'
    arg[0]: a 'MSG'
    stack temp[0]: 460818
    stack temp[1]: a 'ShellView'
{08CD01A0: cf 08CD0181, sp 08CD01B4, bp 08CD019C, ip 19, 'InputState'>>pumpMessage:}
    receiver: a 'InputState'
    arg[0]: a 'MSG'
{08CD0180: cf 08CD0155, sp 08CD0194, bp 08CD0170, ip 18, 'InputState'>>loopWhile:}
    receiver: a 'InputState'
    arg[0]: [] @ 9 in 'InputState'>>mainLoop
    stack temp[0]: a 'MSG'
    stack temp[1]: true
    stack temp[2]: [] @ 8 in 'Toolbar'>>updateSizePosted
{08CD0154: cf 08CD0139, sp 08CD0168, bp 08CD0154, ip 15, 'InputState'>>mainLoop}
    receiver: a 'InputState'
{08CD0138: cf 08CD011D, sp 08CD014C, bp 08CD0138, ip 14, [] in 'InputState'>>forkMain}
    receiver: a 'InputState'
{08CD011C: cf 08CD0101, sp 08CD0130, bp 08CD011C, ip 13, 'ExceptionHandler'('ExceptionHandlerAbstract')>>markAndTry}
    receiver: a 'ExceptionHandler'
{08CD0100: cf 08CD00D9, sp 08CD0114, bp 08CD00FC, ip 22, [] in 'ExceptionHandler'('ExceptionHandlerAbstract')>>try:}
    receiver: a 'ExceptionHandler'
{08CD00D8: cf 08CD00B9, sp 08CD00EC, bp 08CD00D4, ip 18, 'BlockClosure'>>ifCurtailed:}
    receiver: [] @ 0 in nil
    arg[0]: [] @ 39 in 'ExceptionHandlerAbstract'>>try:
{08CD00B8: cf 08CD0095, sp 08CD00CC, bp 08CD00B0, ip 3, 'BlockClosure'>>ensure:}
    receiver: [] @ 17 in 'ExceptionHandlerAbstract'>>try:
    arg[0]: [] @ 39 in 'ExceptionHandlerAbstract'>>try:
    stack temp[0]: nil
{08CD0094: cf 08CD006D, sp 08CD00A8, bp 08CD0088, ip 44, 'ExceptionHandler'('ExceptionHandlerAbstract')>>try:}
    receiver: a 'ExceptionHandler'
    arg[0]: [] @ 10 in 'InputState'>>forkMain
    stack temp[0]: nil
    stack temp[1]: a 'Process'('Main' base 08CD0000 [ACTIVE] in 'SessionManager'>>quit: sp=00000000 ip=3 list=10050010)
    env temp[0]: nil
{08CD006C: cf 08CD0049, sp 08CD0080, bp 08CD0064, ip 7, 'BlockClosure'>>on:do:}
    receiver: [] @ 10 in 'InputState'>>forkMain
    arg[0]: 'ProcessTermination'
    arg[1]: [] @ 13 in 'BlockClosure'>>newProcess
{08CD0048: cf 00000001, sp 08CD005C, bp 08CD0048, ip 17, [] in 'BlockClosure'>>newProcess}
    receiver: [] @ 10 in 'InputState'>>forkMain
<Bottom of stack>
*----> Stack <----*
[08CD0674: 397]-->0
[08CD0670: 396]-->a 'DevelopmentSessionManager'
[08CD066C: 395]-->73827114
```

2) Evaluate the same as 1) but inside a `#deferredValue` block
Evaluate: `[SessionManager current quit: 0] deferredValue`
Result:
Windows Server 2012 and Windows 10: Dolphin exits cleanly without warns nor errors file

3) Patch `SessionManager>>quit:`
```
!SessionManager methodsFor!
quit: anInteger
    "Private - Force a close down of the session with the specified exit code."

    self onExit == true ifTrue: [[self primQuit: anInteger] deferredValue].
    ^false    "cancelled" ! ! 
```
We thought it would work but again the crash in windows 2012 and the stack (even with deferredValue) is:
```
********************************************************************************
************************** Dolphin Crash Dump Report ***************************
18:05:27 p.m., 21/04/2017: Zafiro.exe caused an unhandled Win32 Exception A0000204
at 74F15B68 in module 74F00000 (C:\Windows\SYSTEM32\KERNELBASE.dll)
*----> Exception Parameters <----*
00000000    
*----> CPU Context for thread 0x78a8 <----*
EAX = 0018E420 EBX = 244302A4 ECX = 00000001
ESI = 244302C4 EDI = 208F50E7 EIP = 74F15B68
ESP = 0018E420 EBP = 0018E478 EFL = 00000212
CS = 0023 SS = 002B DS = 002B
ES = 002B FS = 0053 GS = 002B
*----> VM Context <----*
Process: {24430004:size 1023 words, suspended frame 24430049, priority 5, callbacks 0
last failure 0:nil, FPE mask 3, thread 00000000}
Active Method: SessionManager>>quit:
IP: 208F50E7 (15)
SP: 244302C4
BP: 244302A4 (153)
ActiveFrame: {244302A8: cf 24430289, sp 00000000, bp 244302A4, ip 0, [] in ZafiroSessionManager(SessionManager)>>quit:}
    receiver: a ZafiroSessionManager
New Method: SessionManager>>primQuit:
Message Selector: primQuit:
*----> Stack Back Trace <----*
{244302A8: cf 24430289, sp 00000000, bp 244302A4, ip 0, [] in ZafiroSessionManager(SessionManager)>>quit:}
    receiver: a ZafiroSessionManager
{24430288: cf 24430261, sp 2443029C, bp 24430280, ip 4, DeferredValue>>_evaluate:withArguments:}
    receiver: a DeferredValue
    arg[0]: [] @ 14 in SessionManager>>quit:
    arg[1]: a Array
{24430260: cf 2443023D, sp 24430278, bp 24430258, ip 24, [] in DeferredValue>>evaluate:withArguments:at:}
    receiver: a DeferredValue
{2443023C: cf 24430221, sp 24430250, bp 2443023C, ip 13, ExceptionHandlerSet(ExceptionHandlerAbstract)>>markAndTry}
    receiver: a ExceptionHandlerSet
{24430220: cf 244301F9, sp 24430234, bp 2443021C, ip 22, [] in ExceptionHandlerSet(ExceptionHandlerAbstract)>>try:}
    receiver: a ExceptionHandlerSet
{244301F8: cf 244301D9, sp 2443020C, bp 244301F4, ip 18, BlockClosure>>ifCurtailed:}
    receiver: [] @ 0 in nil
    arg[0]: [] @ 39 in ExceptionHandlerAbstract>>try:
{244301D8: cf 244301B5, sp 244301EC, bp 244301D0, ip 3, BlockClosure>>ensure:}
    receiver: [] @ 17 in ExceptionHandlerAbstract>>try:
    arg[0]: [] @ 39 in ExceptionHandlerAbstract>>try:
    stack temp[0]: nil
{244301B4: cf 2443018D, sp 244301C8, bp 244301A8, ip 44, ExceptionHandlerSet(ExceptionHandlerAbstract)>>try:}
    receiver: a ExceptionHandlerSet
    arg[0]: [] @ 20 in DeferredValue>>evaluate:withArguments:at:
    stack temp[0]: a ExceptionHandler
    stack temp[1]: a Process(DeferredValue>>evaluate:withArguments:at: base 24430000 [ACTIVE] in SessionManager>>quit: sp=00000000 ip=0 list=032B0010)
    env temp[0]: nil
{2443018C: cf 2443016D, sp 244301A0, bp 24430188, ip 4, BlockClosure>>onDo:}
    receiver: [] @ 20 in DeferredValue>>evaluate:withArguments:at:
    arg[0]: a ExceptionHandlerSet
{2443016C: cf 24430141, sp 24430180, bp 2443015C, ip 13, BlockClosure>>on:do:on:do:}
    receiver: [] @ 20 in DeferredValue>>evaluate:withArguments:at:
    arg[0]: Error
    arg[1]: [] @ 34 in DeferredValue>>evaluate:withArguments:at:
    arg[2]: ProcessTermination
    arg[3]: [] @ 48 in DeferredValue>>evaluate:withArguments:at:
{24430140: cf 2443011D, sp 24430154, bp 24430138, ip 61, [] in DeferredValue>>evaluate:withArguments:at:}
    receiver: a DeferredValue
{2443011C: cf 24430101, sp 24430130, bp 2443011C, ip 13, ExceptionHandler(ExceptionHandlerAbstract)>>markAndTry}
    receiver: a ExceptionHandler
{24430100: cf 244300D9, sp 24430114, bp 244300FC, ip 22, [] in ExceptionHandler(ExceptionHandlerAbstract)>>try:}
    receiver: a ExceptionHandler
{244300D8: cf 244300B9, sp 244300EC, bp 244300D4, ip 18, BlockClosure>>ifCurtailed:}
    receiver: [] @ 0 in nil
    arg[0]: [] @ 39 in ExceptionHandlerAbstract>>try:
{244300B8: cf 24430095, sp 244300CC, bp 244300B0, ip 3, BlockClosure>>ensure:}
    receiver: [] @ 17 in ExceptionHandlerAbstract>>try:
    arg[0]: [] @ 39 in ExceptionHandlerAbstract>>try:
    stack temp[0]: nil
{24430094: cf 2443006D, sp 244300A8, bp 24430088, ip 44, ExceptionHandler(ExceptionHandlerAbstract)>>try:}
    receiver: a ExceptionHandler
    arg[0]: [] @ 11 in DeferredValue>>evaluate:withArguments:at:
    stack temp[0]: nil
    stack temp[1]: a Process(DeferredValue>>evaluate:withArguments:at: base 24430000 [ACTIVE] in SessionManager>>quit: sp=00000000 ip=0 list=032B0010)
    env temp[0]: nil
{2443006C: cf 24430049, sp 24430080, bp 24430064, ip 7, BlockClosure>>on:do:}
    receiver: [] @ 11 in DeferredValue>>evaluate:withArguments:at:
    arg[0]: ProcessTermination
    arg[1]: [] @ 13 in BlockClosure>>newProcess
{24430048: cf 00000001, sp 2443005C, bp 24430048, ip 17, [] in BlockClosure>>newProcess}
    receiver: [] @ 11 in DeferredValue>>evaluate:withArguments:at:
<Bottom of stack>
*----> Stack <----*
[244302C4: 161]-->0
[244302C0: 160]-->a ZafiroSessionManager
[244302BC: 159]-->304185682
[244302B8: 158]-->0
[244302B4: 157]-->0
[244302B0: 156]-->304185668
[244302AC: 155]-->[] @ 14 in SessionManager>>quit:
[244302A8: 154]-->SessionManager>>quit:
[244302A4: 153]-->0
[244302A0: 152]-->a ZafiroSessionManager
[2443029C: 151]-->304185664
[24430298: 150]-->304185678
[24430294: 149]-->3
[24430290: 148]-->304185648
[2443028C: 147]-->0
[24430288: 146]-->DeferredValue>>_evaluate:withArguments:
[24430284: 145]-->a Array
[24430280: 144]-->[] @ 14 in SessionManager>>quit:
[2443027C: 143]-->a DeferredValue
[24430278: 142]-->a DeferredValue
[24430274: 141]-->304185644
[24430270: 140]-->304185660
[2443026C: 139]-->23
[24430268: 138]-->304185630
[24430264: 137]-->[] @ 20 in DeferredValue>>evaluate:withArguments:at:
[24430260: 136]-->DeferredValue>>evaluate:withArguments:at:
[2443025C: 135]-->[] @ 14 in SessionManager>>quit:
[24430258: 134]-->a Array
[24430254: 133]-->a DeferredValue
[24430250: 132]-->304185630
[2443024C: 131]-->304185640
[24430248: 130]-->12
[24430244: 129]-->304185616
[24430240: 128]-->a Context for: nil frame: 2443023d
[2443023C: 127]-->ExceptionHandlerAbstract>>markAndTry
[24430238: 126]-->a ExceptionHandlerSet
[24430234: 125]-->304185614
[24430230: 124]-->304185626
[2443022C: 123]-->21
[24430228: 122]-->304185596
[24430224: 121]-->[] @ 17 in ExceptionHandlerAbstract>>try:
[24430220: 120]-->ExceptionHandlerAbstract>>try:
[2443021C: 119]-->a Process(DeferredValue>>evaluate:withArguments:at: base 24430000 [ACTIVE] in SessionManager>>quit: sp=00000000 ip=0 list=032B0010)
[24430218: 118]-->a ExceptionHandlerSet
[24430214: 117]-->[] @ 9 in BlockClosure>>ifCurtailed:
[24430210: 116]-->[] @ 17 in ExceptionHandlerAbstract>>try:
[2443020C: 115]-->304185594
[24430208: 114]-->304185606
[24430204: 113]-->17
[24430200: 112]-->304185580
[244301FC: 111]-->0
[244301F8: 110]-->BlockClosure>>ifCurtailed:
[244301F4: 109]-->[] @ 39 in ExceptionHandlerAbstract>>try:
[244301F0: 108]-->[] @ 0 in nil
[244301EC: 107]-->304185576
[244301E8: 106]-->304185590
[244301E4: 105]-->2
[244301E0: 104]-->304185562
[244301DC: 103]-->0
[244301D8: 102]-->BlockClosure>>ensure:
[244301D4: 101]-->nil
[244301D0: 100]-->[] @ 39 in ExceptionHandlerAbstract>>try:
[244301CC: 99]-->[] @ 17 in ExceptionHandlerAbstract>>try:
[244301C8: 98]-->304185556
[244301C4: 97]-->304185572
[244301C0: 96]-->43
[244301BC: 95]-->304185542
[244301B8: 94]-->a Context for: nil frame: 244301b5
[244301B4: 93]-->ExceptionHandlerAbstract>>try:
[244301B0: 92]-->a Process(DeferredValue>>evaluate:withArguments:at: base 24430000 [ACTIVE] in SessionManager>>quit: sp=00000000 ip=0 list=032B0010)
[244301AC: 91]-->a ExceptionHandler
[244301A8: 90]-->[] @ 20 in DeferredValue>>evaluate:withArguments:at:
[244301A4: 89]-->a ExceptionHandlerSet
[244301A0: 88]-->304185540
[2443019C: 87]-->304185552
[24430198: 86]-->3
[24430194: 85]-->304185526
[24430190: 84]-->0
[2443018C: 83]-->BlockClosure>>onDo:
[24430188: 82]-->a ExceptionHandlerSet
[24430184: 81]-->[] @ 20 in DeferredValue>>evaluate:withArguments:at:
[24430180: 80]-->304185518
[2443017C: 79]-->304185536
[24430178: 78]-->12
[24430174: 77]-->304185504
[24430170: 76]-->0
[2443016C: 75]-->BlockClosure>>on:do:on:do:
[24430168: 74]-->[] @ 48 in DeferredValue>>evaluate:withArguments:at:
[24430164: 73]-->ProcessTermination
[24430160: 72]-->[] @ 34 in DeferredValue>>evaluate:withArguments:at:
[2443015C: 71]-->Error
[24430158: 70]-->[] @ 20 in DeferredValue>>evaluate:withArguments:at:
[24430154: 69]-->304185500
[24430150: 68]-->304185514
[2443014C: 67]-->60
[24430148: 66]-->304185486
[24430144: 65]-->[] @ 11 in DeferredValue>>evaluate:withArguments:at:
[24430140: 64]-->DeferredValue>>evaluate:withArguments:at:
[2443013C: 63]-->[] @ 14 in SessionManager>>quit:
[24430138: 62]-->a Array
[24430134: 61]-->a DeferredValue
[24430130: 60]-->304185486
[2443012C: 59]-->304185496
[24430128: 58]-->12
[24430124: 57]-->304185472
[24430120: 56]-->a Context for: nil frame: 2443011d
[2443011C: 55]-->ExceptionHandlerAbstract>>markAndTry
[24430118: 54]-->a ExceptionHandler
[24430114: 53]-->304185470
[24430110: 52]-->304185482
[2443010C: 51]-->21
[24430108: 50]-->304185452
[24430104: 49]-->[] @ 17 in ExceptionHandlerAbstract>>try:
[24430100: 48]-->ExceptionHandlerAbstract>>try:
[244300FC: 47]-->a Process(DeferredValue>>evaluate:withArguments:at: base 24430000 [ACTIVE] in SessionManager>>quit: sp=00000000 ip=0 list=032B0010)
[244300F8: 46]-->a ExceptionHandler
[244300F4: 45]-->[] @ 9 in BlockClosure>>ifCurtailed:
[244300F0: 44]-->[] @ 17 in ExceptionHandlerAbstract>>try:
[244300EC: 43]-->304185450
[244300E8: 42]-->304185462
[244300E4: 41]-->17
[244300E0: 40]-->304185436
[244300DC: 39]-->0
[244300D8: 38]-->BlockClosure>>ifCurtailed:
[244300D4: 37]-->[] @ 39 in ExceptionHandlerAbstract>>try:
[244300D0: 36]-->[] @ 0 in nil
[244300CC: 35]-->304185432
[244300C8: 34]-->304185446
[244300C4: 33]-->2
[244300C0: 32]-->304185418
[244300BC: 31]-->0
[244300B8: 30]-->BlockClosure>>ensure:
[244300B4: 29]-->nil
[244300B0: 28]-->[] @ 39 in ExceptionHandlerAbstract>>try:
[244300AC: 27]-->[] @ 17 in ExceptionHandlerAbstract>>try:
[244300A8: 26]-->304185412
[244300A4: 25]-->304185428
[244300A0: 24]-->43
[2443009C: 23]-->304185398
[24430098: 22]-->a Context for: nil frame: 24430095
[24430094: 21]-->ExceptionHandlerAbstract>>try:
[24430090: 20]-->a Process(DeferredValue>>evaluate:withArguments:at: base 24430000 [ACTIVE] in SessionManager>>quit: sp=00000000 ip=0 list=032B0010)
[2443008C: 19]-->nil
[24430088: 18]-->[] @ 11 in DeferredValue>>evaluate:withArguments:at:
[24430084: 17]-->a ExceptionHandler
[24430080: 16]-->304185394
[2443007C: 15]-->304185408
[24430078: 14]-->6
[24430074: 13]-->304185380
[24430070: 12]-->0
[2443006C: 11]-->BlockClosure>>on:do:
[24430068: 10]-->[] @ 13 in BlockClosure>>newProcess
[24430064: 9]-->ProcessTermination
[24430060: 8]-->[] @ 11 in DeferredValue>>evaluate:withArguments:at:
[2443005C: 7]-->304185380
[24430058: 6]-->304185390
[24430054: 5]-->16
[24430050: 4]-->0
[2443004C: 3]-->[] @ 9 in BlockClosure>>newProcess
[24430048: 2]-->BlockClosure>>newProcess
[24430044: 1]-->[] @ 11 in DeferredValue>>evaluate:withArguments:at:
<Bottom of stack>
***** End of crash report *****
```

**Comment by blairmcg** [comment]
I'm closing this because without a repro it is not actionable. There have also been many related changes since.
